app/lib/link_details_extractor.rb

```diff
@@ -269,14 +269,31 @@ def structured_data
   end
 
   def document
-    @document ||= Nokogiri::HTML(@html, nil, encoding)
+    @document ||= begin
+      encoding = detect_encoding || parse_encoding || header_encoding || 'UTF-8'
+      @document || Nokogiri::HTML(@html, nil, encoding)
+    end
   end
 
-  def encoding
-    @encoding ||= begin
-      guess = detector.detect(@html, @html_charset)
-      guess&.fetch(:confidence, 0).to_i > 60 ? guess&.fetch(:encoding, nil) : nil
-    end
+  def detect_encoding
+    guess = detector.detect(@html, @html_charset)
+    guess&.fetch(:confidence, 0).to_i > 60 ? guess&.fetch(:encoding, nil) : nil
+  end
+
+  def parse_encoding
+    parsed_document = Nokogiri::HTML(@html)
+    source_with_detected_encoding = parsed_document.to_s
+    return unless source_with_detected_encoding.valid_encoding?
```

Comment:
I wonder if there is a less resource-intensive way to check that everything can be decoded correctly… If I understand the issue correctly, the difficult part is that exceptions do not arise when creating `parsed_document` but when accessing specific attributes down the line.

As `LinkDetailsExtractor` is only used in a single place (`FetchLinkCardServices`) and `#to_preview_card_attributes` is called unconditionally in the same method, maybe an alternative is to build the `preview_card_attributes` right from `initialize` and perform the error handling there.

That is, try charsets in order (confident CharlockHolmes/ICU guess, then parsed from document, then header, then utf-8) and stop at the first that raises no error.

Reply:
Ah, you're right, I missed this was a very generic `ArgumentError`, I agree that rescuing on `ArgumentError` is a bit too broad and trying to match on the error text is very awkward.

I'm not too worried about other places raising exceptions, the effective used surface of `LinkDetailsExtractor` consists of only the initializer, `#author_account`, `#canonical_url`, and `#to_preview_card_attributes`. Those could all be computed in the initializer with proper exception handling, and the `@document` thrown away, not to be accessed again.